File: zig/src/compliance.zig
//! Zig 0.16 compliance runner — std.json against the shared JSON catalog.
const std = @import("std");

pub fn main(init: std.process.Init) !void {
    const alloc = init.gpa;
    const io = init.io;

    var json_out: ?[]const u8 = null;
    var formats: std.ArrayList([]const u8) = .empty;
    defer formats.deinit(alloc);

    var it = std.process.Args.Iterator.init(init.minimal.args);
    _ = it.next();
    while (it.next()) |a| {
        if (std.mem.eql(u8, a, "--json-out") or std.mem.eql(u8, a, "-o")) {
            json_out = it.next();
        } else if (std.mem.eql(u8, a, "--format") or std.mem.eql(u8, a, "-f")) {
            if (it.next()) |f| try formats.append(alloc, f);
        }
    }

    const cwd = std.Io.Dir.cwd();
    var root_rel: []const u8 = ".";
    var found = false;
    var up: usize = 0;
    while (up < 8) : (up += 1) {
        const probe = try std.fs.path.join(alloc, &.{ root_rel, "compliance", "data", "json" });
        defer alloc.free(probe);
        if (cwd.openDir(io, probe, .{ .iterate = true })) |d| {
            d.close(io);
            found = true;
            break;
        } else |_| {}
        const next = try std.fs.path.join(alloc, &.{ root_rel, ".." });
        if (!std.mem.eql(u8, root_rel, ".")) alloc.free(root_rel);
        root_rel = next;
    }
    if (!found) return error.NoCatalog;
    defer if (!std.mem.eql(u8, root_rel, ".")) alloc.free(root_rel);

    const data_dir = try std.fs.path.join(alloc, &.{ root_rel, "compliance", "data", "json" });
    defer alloc.free(data_dir);

    var want_json = formats.items.len == 0;
    var want_pb = formats.items.len == 0;
    for (formats.items) |f| {
        if (std.ascii.eqlIgnoreCase(f, "json")) want_json = true;
        if (std.ascii.eqlIgnoreCase(f, "protobuf")) want_pb = true;
    }

    var results: std.ArrayList(u8) = .empty;
    defer results.deinit(alloc);
    try results.appendSlice(alloc, "[");
    var first = true;
    var p: usize = 0;
    var f: usize = 0;
    var total: usize = 0;

    if (want_json) {
        var dir = try cwd.openDir(io, data_dir, .{ .iterate = true });
        defer dir.close(io);
        var walker = dir.iterate();
        while (try walker.next(io)) |ent| {
            if (ent.kind != .file or !std.mem.endsWith(u8, ent.name, ".json")) continue;
            if (ent.name.len > 0 and ent.name[0] == '_') continue;
            const raw = try readFile(alloc, io, data_dir, ent.name);
            defer alloc.free(raw);
            const parsed = std.json.parseFromSlice(std.json.Value, alloc, raw, .{}) catch continue;
            defer parsed.deinit();
            const suite = parsed.value;
            const cases = (suite.object.get("cases") orelse continue).array;
            const standard = suite.object.get("standard").?.string;
            const version = suite.object.get("version").?.string;
            const standard_url = if (suite.object.get("standard_url")) |u| u.string else "";
            for (cases.items) |c| {
                const id = c.object.get("id").?.string;
                const expect = c.object.get("expect").?.string;
                const input = c.object.get("input").?.string;
                const enc = if (c.object.get("input_encoding")) |e| e.string else "utf-8";
                if (!std.mem.eql(u8, enc, "utf-8") and enc.len != 0) continue;
                if (tooDeep(input)) continue;
                total += 1;
                var outcome: []const u8 = "pass";
                var observed: []const u8 = "ok";
                if (std.json.parseFromSlice(std.json.Value, alloc, input, .{})) |ok| {
                    ok.deinit();
                    if (std.mem.eql(u8, expect, "reject")) {
                        outcome = "fail";
                        observed = "accepted";
                        f += 1;
                    } else p += 1;
                } else |_| {
                    if (std.mem.eql(u8, expect, "reject") or std.mem.eql(u8, expect, "any")) {
                        p += 1;
                        observed = "rejected";
                    } else {
                        outcome = "fail";
                        observed = "rejected";
                        f += 1;
                    }
                }
                if (!first) try results.appendSlice(alloc, ",");
                first = false;
                const qid = try std.json.Stringify.valueAlloc(alloc, id, .{});
                defer alloc.free(qid);
                const qstd = try std.json.Stringify.valueAlloc(alloc, standard, .{});
                defer alloc.free(qstd);
                const qurl = try std.json.Stringify.valueAlloc(alloc, standard_url, .{});
                defer alloc.free(qurl);
                const qver = try std.json.Stringify.valueAlloc(alloc, version, .{});
                defer alloc.free(qver);
                const qreq = try std.json.Stringify.valueAlloc(alloc, c.object.get("requirement").?.string, .{});
                defer alloc.free(qreq);
                const qexp = try std.json.Stringify.valueAlloc(alloc, expect, .{});
                defer alloc.free(qexp);
                const qsec = try std.json.Stringify.valueAlloc(alloc, c.object.get("section_url").?.string, .{});
                defer alloc.free(qsec);
                const qobs = try std.json.Stringify.valueAlloc(alloc, observed, .{});
                defer alloc.free(qobs);
                const qout = try std.json.Stringify.valueAlloc(alloc, outcome, .{});
                defer alloc.free(qout);
                const row = try std.fmt.allocPrint(
                    alloc,
                    "{{\"id\":{s},\"language\":\"zig\",\"serializer\":\"std.json\",\"serializer_version\":\"\",\"format\":\"json\",\"standard\":{s},\"standard_url\":{s},\"version\":{s},\"version_key\":\"json.{s}\",\"requirement\":{s},\"expect\":{s},\"section\":\"\",\"section_title\":\"\",\"section_url\":{s},\"paragraph\":\"\",\"title\":\"\",\"input\":\"\",\"input_encoding\":\"utf-8\",\"detail\":\"\",\"observed\":{s},\"outcome\":{s}}}",
                    .{ qid, qstd, qurl, qver, version, qreq, qexp, qsec, qobs, qout },
                );
                defer alloc.free(row);
                try results.appendSlice(alloc, row);
            }
        }
    }
    if (want_pb) {
        const pb_dir = try std.fs.path.join(alloc, &.{ root_rel, "compliance", "data", "protobuf" });
        defer alloc.free(pb_dir);
        if (cwd.openDir(io, pb_dir, .{ .iterate = true })) |dir| {
            defer dir.close(io);
            var walker = dir.iterate();
            while (try walker.next(io)) |ent| {
                if (ent.kind != .file or !std.mem.endsWith(u8, ent.name, ".json")) continue;
                if (ent.name.len > 0 and ent.name[0] == '_') continue;
                const raw = try readFile(alloc, io, pb_dir, ent.name);
                defer alloc.free(raw);
                const parsed = std.json.parseFromSlice(std.json.Value, alloc, raw, .{}) catch continue;
                defer parsed.deinit();
                const suite = parsed.value;
                const cases = (suite.object.get("cases") orelse continue).array;
                const standard = suite.object.get("standard").?.string;
                const version = suite.object.get("version").?.string;
                const standard_url = if (suite.object.get("standard_url")) |u| u.string else "";
                for (cases.items) |c| {
                    const id = c.object.get("id").?.string;
                    const expect = c.object.get("expect").?.string;
                    const input = c.object.get("input").?.string;
                    const enc = if (c.object.get("input_encoding")) |e| e.string else "utf-8";
                    const schema = if (c.object.get("schema")) |s| s.string else "";
                    total += 1;
                    var outcome: []const u8 = "pass";
                    var observed: []const u8 = "ok";
                    const bytes = inputBytes(alloc, input, enc) catch {
                        if (std.mem.eql(u8, expect, "reject") or std.mem.eql(u8, expect, "any")) {
                            p += 1;
                            observed = "rejected";
                        } else {
                            outcome = "fail";
                            observed = "rejected";
                            f += 1;
                        }
                        try appendPbRow(alloc, &results, &first, id, standard, standard_url, version, c, expect, observed, outcome);
                        continue;
                    };
                    defer alloc.free(bytes);
                    if (decodeProtobuf(alloc, bytes, schema)) |_| {
                        if (std.mem.eql(u8, expect, "reject")) {
                            outcome = "fail";
                            observed = "accepted";
                            f += 1;
                        } else p += 1;
                    } else |_| {
                        if (std.mem.eql(u8, expect, "reject") or std.mem.eql(u8, expect, "any")) {
                            p += 1;
                            observed = "rejected";
                        } else {
                            outcome = "fail";
                            observed = "rejected";
                            f += 1;
                        }
                    }
                    try appendPbRow(alloc, &results, &first, id, standard, standard_url, version, c, expect, observed, outcome);
                }
            }
        } else |_| {}
    }
    try results.appendSlice(alloc, "]");
    std.debug.print("Serialization compliance (library deviations are catalogued, not a red build)\n", .{});
    std.debug.print("  {d} pass  {d} fail  0 skip  0 error  {d} total\n", .{ p, f, total });
    if (json_out) |out_path| {
        const doc = try std.fmt.allocPrint(
            alloc,
            "{{\"schema\":\"gld.dashboard.compliance/1\",\"generated_at\":\"\",\"language\":\"zig\",\"languages\":[\"zig\"],\"policy\":\"report-only\",\"scope\":{{\"formats\":[\"json\",\"protobuf\"]}},\"passed\":{d},\"failed\":{d},\"skipped\":0,\"errors\":0,\"catalog_errors\":[],\"serializer_errors\":[],\"results\":{s}}}\n",
            .{ p, f, results.items },
        );
        defer alloc.free(doc);
        if (std.fs.path.dirname(out_path)) |d| {
            cwd.createDirPath(io, d) catch {};
        }
        try cwd.writeFile(io, .{ .sub_path = out_path, .data = doc });
        std.debug.print("\nWrote {s}\n", .{out_path});
    }
}

fn appendPbRow(alloc: std.mem.Allocator, results: *std.ArrayList(u8), first: *bool, id: []const u8, standard: []const u8, standard_url: []const u8, version: []const u8, c: std.json.Value, expect: []const u8, observed: []const u8, outcome: []const u8) !void {
    if (!first.*) try results.appendSlice(alloc, ",");
    first.* = false;
    const qid = try std.json.Stringify.valueAlloc(alloc, id, .{});
    defer alloc.free(qid);
    const qstd = try std.json.Stringify.valueAlloc(alloc, standard, .{});
    defer alloc.free(qstd);
    const qurl = try std.json.Stringify.valueAlloc(alloc, standard_url, .{});
    defer alloc.free(qurl);
    const qver = try std.json.Stringify.valueAlloc(alloc, version, .{});
    defer alloc.free(qver);
    const qreq = try std.json.Stringify.valueAlloc(alloc, c.object.get("requirement").?.string, .{});
    defer alloc.free(qreq);
    const qexp = try std.json.Stringify.valueAlloc(alloc, expect, .{});
    defer alloc.free(qexp);
    const qsec = try std.json.Stringify.valueAlloc(alloc, c.object.get("section_url").?.string, .{});
    defer alloc.free(qsec);
    const qobs = try std.json.Stringify.valueAlloc(alloc, observed, .{});
    defer alloc.free(qobs);
    const qout = try std.json.Stringify.valueAlloc(alloc, outcome, .{});
    defer alloc.free(qout);
    const row = try std.fmt.allocPrint(
        alloc,
        "{{\"id\":{s},\"language\":\"zig\",\"serializer\":\"protobuf-wire\",\"serializer_version\":\"\",\"format\":\"protobuf\",\"standard\":{s},\"standard_url\":{s},\"version\":{s},\"version_key\":\"protobuf.{s}\",\"requirement\":{s},\"expect\":{s},\"section\":\"\",\"section_title\":\"\",\"section_url\":{s},\"paragraph\":\"\",\"title\":\"\",\"input\":\"\",\"input_encoding\":\"\",\"detail\":\"\",\"observed\":{s},\"outcome\":{s}}}",
        .{ qid, qstd, qurl, qver, version, qreq, qexp, qsec, qobs, qout },
    );
    defer alloc.free(row);
    try results.appendSlice(alloc, row);
}

fn inputBytes(alloc: std.mem.Allocator, input: []const u8, enc: []const u8) ![]u8 {
    if (std.mem.eql(u8, enc, "hex")) {
        var n: usize = 0;
        for (input) |ch| {
            if (ch != ' ' and ch != '\n' and ch != '\t' and ch != '\r') n += 1;
        }
        if (n == 0) return alloc.alloc(u8, 0);
        if (n % 2 != 0) return error.OddHex;
        const out = try alloc.alloc(u8, n / 2);
        var i: usize = 0;
        var o: usize = 0;
        while (i < input.len) : (i += 1) {
            const ch = input[i];
            if (ch == ' ' or ch == '\n' or ch == '\t' or ch == '\r') continue;
            const hi = hexNibble(ch) orelse return error.BadHex;
            i += 1;
            while (i < input.len and (input[i] == ' ' or input[i] == '\n')) i += 1;
            if (i >= input.len) return error.OddHex;
            const lo = hexNibble(input[i]) orelse return error.BadHex;
            out[o] = (hi << 4) | lo;
            o += 1;
        }
        return out;
    }
    return alloc.dupe(u8, input);
}

fn hexNibble(c: u8) ?u8 {
    return switch (c) {
        '0'...'9' => c - '0',
        'a'...'f' => c - 'a' + 10,
        'A'...'F' => c - 'A' + 10,
        else => null,
    };
}

fn pbVarint(data: []const u8, i: *usize) !u64 {
    var r: u64 = 0;
    var shift: u6 = 0;
    while (i.* < data.len) {
        const b = data[i.*];
        i.* += 1;
        r |= @as(u64, b & 0x7f) << shift;
        if (b & 0x80 == 0) return r;
        if (shift >= 63) return error.VarintTooLong;
        shift += 7;
    }
    return error.TruncatedVarint;
}

fn decodeProtobuf(alloc: std.mem.Allocator, data: []const u8, schema: []const u8) !void {
    if (std.mem.eql(u8, schema, "json")) {
        const parsed = std.json.parseFromSlice(std.json.Value, alloc, data, .{}) catch return error.BadJSON;
        defer parsed.deinit();
        const obj = switch (parsed.value) {
            .object => |o| o,
            else => return error.NotObject,
        };
        if (obj.get("n")) |n| {
            switch (n) {
                .integer => {},
                .string => |s| _ = std.fmt.parseInt(i32, s, 10) catch return error.BadType,
                .null => {},
                else => return error.BadType,
            }
        }
        return;
    }
    var i: usize = 0;
    while (i < data.len) {
        const key = try pbVarint(data, &i);
        const wt = key & 7;
        if (wt == 0) {
            _ = try pbVarint(data, &i);
        } else if (wt == 1) {
            if (i + 8 > data.len) return error.Truncated;
            i += 8;
        } else if (wt == 5) {
            if (i + 4 > data.len) return error.Truncated;
            i += 4;
        } else if (wt == 2) {
            const n = try pbVarint(data, &i);
            if (i + n > data.len) return error.Truncated;
            i += @intCast(n);
        } else return error.BadWire;
    }
}

fn tooDeep(s: []const u8) bool {
    if (s.len > 200_000) return true;
    var n: usize = 0;
    for (s) |c| {
        if (c == '[' or c == '{') {
            n += 1;
            if (n > 4000) return true;
        }
    }
    return false;
}

fn readFile(alloc: std.mem.Allocator, io: std.Io, dir: []const u8, name: []const u8) ![]u8 {
    const path = try std.fs.path.join(alloc, &.{ dir, name });
    defer alloc.free(path);
    const file = try std.Io.Dir.cwd().openFile(io, path, .{});
    defer file.close(io);
    var buf: [4096]u8 = undefined;
    var reader = file.reader(io, &buf);
    return reader.interface.allocRemaining(alloc, .unlimited);
}

File: javascript/src/compliance.mjs
#!/usr/bin/env node
/**
 * JavaScript compliance runner — same catalog as Python.
 *
 *   node javascript/src/compliance.mjs --json-out logs/compliance/javascript.json
 */
import { readdirSync, readFileSync, mkdirSync, writeFileSync } from 'node:fs';
import { dirname, join, resolve } from 'node:path';
import { fileURLToPath } from 'node:url';
import { createRequire } from 'node:module';
import { toObject as flexToObject } from 'flatbuffers/mjs/flexbuffers.js';

const require = createRequire(import.meta.url);
const here = dirname(fileURLToPath(import.meta.url));
const repoRoot = resolve(here, '../..');
const catalogRoot = join(repoRoot, 'compliance/data');

function pkgVersion(name) {
  if (!name) return '';
  const files = [
    join(here, '../node_modules', name, 'package.json'),
    join(repoRoot, 'javascript/node_modules', name, 'package.json'),
  ];
  for (const p of files) {
    try {
      const v = JSON.parse(readFileSync(p, 'utf8')).version;
      if (v) return String(v);
    } catch {
      /* try the next resolver */
    }
  }
  try {
    return require(`${name}/package.json`).version || '';
  } catch {
    return '';
  }
}

function parseArgs(argv) {
  const out = { jsonOut: null, formats: [], serializers: [] };
  for (let i = 2; i < argv.length; i++) {
    const a = argv[i];
    if ((a === '--json-out' || a === '-o') && argv[i + 1]) out.jsonOut = argv[++i];
    else if ((a === '--format' || a === '-f') && argv[i + 1]) out.formats.push(argv[++i]);
    else if ((a === '--serializer' || a === '-s') && argv[i + 1]) out.serializers.push(argv[++i]);
  }
  return out;
}

function decodeProtobufJs(protobuf, buf, schema) {
  if (schema === 'json') {
    const v = JSON.parse(buf.toString('utf8'));
    if (v === null || typeof v !== 'object' || Array.isArray(v)) {
      throw new Error('proto3 JSON message must be an object');
    }
    const doc = { n: 0, s: '', ok: false, tags: [] };
    if (v.n != null) doc.n = pbInt32(v.n);
    if (v.s != null) {
      if (typeof v.s !== 'string') throw new Error('s must be a string');
      doc.s = v.s;
    }
    if (v.ok != null) {
      if (typeof v.ok !== 'boolean') throw new Error('ok must be a bool');
      doc.ok = v.ok;
    }
    if (v.tags != null) {
      if (!Array.isArray(v.tags)) throw new Error('tags must be an array');
      doc.tags = v.tags.map(pbInt32);
    }
    return doc;
  }
  const r = protobuf.Reader.create(buf);
  const doc = { n: 0, s: '', ok: false, tags: [] };
  while (r.pos < r.len) {
    const tag = r.uint32();
    const field = tag >>> 3;
    const wt = tag & 7;
    if (field === 1 && wt === 0) doc.n = r.int32();
    else if (field === 2 && wt === 2) {
      const n = r.uint32();
      if (r.pos + n > r.len) throw new Error('truncated length-delimited');
      doc.s = Buffer.from(r.buf.subarray(r.pos, r.pos + n)).toString('utf8');
      r.pos += n;
    } else if (field === 3 && wt === 0) doc.ok = r.bool();
    else if (field === 4 && wt === 0) doc.tags.push(r.int32());
    else if (field === 4 && wt === 2) {
      const n = r.uint32();
      if (r.pos + n > r.len) throw new Error('truncated length-delimited');
      const end = r.pos + n;
      while (r.pos < end) doc.tags.push(r.int32());
    } else {
      r.skipType(wt);
    }
  }
  return doc;
}

function pbInt32(v) {
  if (typeof v === 'number' && Number.isFinite(v)) return v | 0;
  if (typeof v === 'string' && /^-?\d+$/.test(v)) return Number(v);
  throw new Error('int32 must be a number or digit string');
}

function inputBytes(c) {
  if (c.input_encoding === 'hex') {
    const compact = String(c.input || '').replace(/\s+/g, '');
    return Buffer.from(compact, 'hex');
  }
  if (c.input_encoding === 'latin-1') return Buffer.from(String(c.input || ''), 'latin1');
  return Buffer.from(String(c.input ?? ''), 'utf8');
}

function valuesEqual(expected, observed) {
  if (expected && typeof expected === 'object' && !Array.isArray(expected) && Object.keys(expected).length === 1) {
    if ('$hex' in expected) {
      if (!Buffer.isBuffer(observed) && !(observed instanceof Uint8Array)) return false;
      return Buffer.from(observed).equals(Buffer.from(String(expected.$hex).replace(/\s+/g, ''), 'hex'));
    }
    if ('$float' in expected) {
      const t = String(expected.$float).toLowerCase();
      if (t === 'nan') return Number.isNaN(observed);
      if (t === 'inf') return observed === Infinity;
      if (t === '-inf') return observed === -Infinity;
    }
  }
  if (expected === null) return observed === null || observed === undefined;
  if (typeof expected === 'boolean' || typeof observed === 'boolean') return expected === observed;
  if (typeof expected === 'number' && typeof observed === 'number') {
    if (Number.isNaN(expected)) return Number.isNaN(observed);
    return expected === observed;
  }
  if (typeof expected === 'string') {
    if (Buffer.isBuffer(observed) || observed instanceof Uint8Array) {
      try {
        return Buffer.from(observed).toString('utf8') === expected;
      } catch {
        return false;
      }
    }
    return expected === observed;
  }
  if (Array.isArray(expected)) {
    if (!Array.isArray(observed) || expected.length !== observed.length) return false;
    return expected.every((v, i) => valuesEqual(v, observed[i]));
  }
  if (expected && typeof expected === 'object') {
    if (!observed || typeof observed !== 'object' || Array.isArray(observed)) return false;
    const ek = Object.keys(expected);
    const ok = Object.keys(observed);
    if (ek.length !== ok.length) return false;
    return ek.every((k) => valuesEqual(expected[k], observed[k]));
  }
  return expected === observed;
}

function preview(value, limit = 120) {
  let text;
  try {
    if (value === undefined) text = 'undefined';
    else if (typeof value === 'symbol') text = value.toString();
    else text = JSON.stringify(value);
    if (text == null) text = String(value);
  } catch {
    text = String(value);
  }
  if (text.length > limit) return `${text.slice(0, limit - 3)}...`;
  return text;
}

function loadSuites() {
  const suites = [];
  for (const fmt of readdirSync(catalogRoot).sort()) {
    const dir = join(catalogRoot, fmt);
    let files;
    try {
      files = readdirSync(dir).filter((n) => n.endsWith('.json') && !n.startsWith('_'));
    } catch {
      continue;
    }
    for (const name of files.sort()) {
      const path = join(dir, name);
      const raw = JSON.parse(readFileSync(path, 'utf8'));
      if (!raw.cases?.length) continue;
      suites.push({ ...raw, source_path: path });
    }
  }
  if (!suites.length) throw new Error(`No compliance suites under ${catalogRoot}`);
  return suites;
}

function makeAdapters() {
  const adapters = [];

  const add = (name, format, decode, notes, pkg) => {
    adapters.push({ name, format, decode, notes, version: pkgVersion(pkg) });
  };

  add('JSON.parse', 'json', (buf) => JSON.parse(buf.toString('utf8')), 'Node JSON.parse', '');
  adapters[adapters.length - 1].version = `node-${process.versions.node}`;
  try {
    const yaml = require('js-yaml');
    add('js-yaml', 'yaml', (buf) => yaml.load(buf.toString('utf8')), 'js-yaml load', 'js-yaml');
  } catch { /* optional */ }
  try {
    const { Decoder } = require('cbor-x');
    const dec = new Decoder();
    add('cbor-x', 'cbor', (buf) => dec.decode(buf), 'cbor-x Decoder', 'cbor-x');
  } catch { /* optional */ }
  try {
    const cbor = require('cbor');
    add('cbor', 'cbor', (buf) => cbor.decodeFirstSync(buf), 'node-cbor decodeFirstSync', 'cbor');
  } catch { /* optional */ }
  try {
    const { Unpackr } = require('msgpackr');
    const unpackr = new Unpackr({ useRecords: false });
    add('msgpackr', 'msgpack', (buf) => unpackr.unpack(buf), 'msgpackr Unpackr', 'msgpackr');
  } catch { /* optional */ }
  try {
    const mp = require('@msgpack/msgpack');
    add('@msgpack/msgpack', 'msgpack', (buf) => mp.decode(buf), '@msgpack/msgpack decode', '@msgpack/msgpack');
  } catch { /* optional */ }
  try {
    const { BSON } = require('bson');
    add('bson', 'bson', (buf) => BSON.deserialize(buf), 'mongodb bson', 'bson');
  } catch { /* optional */ }
  add('flexbuffers', 'flatbuffers', (buf) => flexToObject(buf), 'flatbuffers flexbuffers', 'flatbuffers');
  try {
    const avro = require('avsc');
    add(
      'avsc',
      'avro',
      (buf, schema) => avro.Type.forSchema(schema ?? 'int').fromBuffer(buf),
      'avsc schemaless',
      'avsc',
    );
  } catch { /* optional */ }
  try {
    const protobuf = require('protobufjs');
    add(
      'protobufjs',
      'protobuf',
      (buf, schema) => decodeProtobufJs(protobuf, buf, schema),
      'protobufjs Reader + proto3 JSON mapping on cmp.Doc',
      'protobufjs',
    );
  } catch { /* optional */ }
  try {
    const bebop = require('bebop');
    add('bebop', 'bebop', (buf) => {
      if (buf.length === 0) return {};
      if (buf.length === 1) return buf[0] !== 0;
      throw new Error('no generated Bebop type for this payload');
    }, 'bebop runtime (bool / empty only)', 'bebop');
  } catch { /* optional */ }

  return adapters.filter((a) => typeof a.decode === 'function');
}

function runOne(suite, c, adapter) {
  const raw = inputBytes(c);
  let observed;
  let decodedOk = true;
  let decodeError = null;
  try {
    observed = adapter.decode(raw, c.schema);
  } catch (err) {
    decodedOk = false;
    decodeError = `${err?.name || 'Error'}: ${err?.message || err}`;
  }
  const base = {
    id: c.id,
    language: 'javascript',
    serializer: adapter.name,
    serializer_version: adapter.version,
    format: suite.format,
    standard: suite.standard,
    standard_url: suite.standard_url || '',
    version: String(suite.version),
    version_key: `${suite.format}.${suite.version}`,
    requirement: c.requirement,
    expect: c.expect,
    section: c.section,
    section_title: c.section_title,
    section_url: c.section_url,
    paragraph: c.paragraph,
    title: c.title,
    input: c.input,
    input_encoding: c.input_encoding || 'utf-8',
    detail: '',
    observed: '',
    outcome: 'pass',
  };
  if (c.expect === 'any') {
    return { ...base, outcome: 'pass', observed: decodeError || preview(observed), detail: 'implementation-defined (recorded, not scored)' };
  }
  if (c.expect === 'reject') {
    if (!decodedOk) return { ...base, outcome: 'pass', observed: decodeError || 'rejected' };
    return { ...base, outcome: 'fail', detail: 'parser accepted input the spec requires to be rejected', observed: `accepted as ${preview(observed)}` };
  }
  if (!decodedOk) {
    return { ...base, outcome: 'fail', detail: 'parser rejected input the spec requires to accept', observed: decodeError || 'rejected' };
  }
  if (
    Object.prototype.hasOwnProperty.call(c, 'decoded') &&
    observed !== Symbol.for('wire-ok') &&
    !valuesEqual(c.decoded, observed)
  ) {
    return { ...base, outcome: 'fail', detail: 'decoded value does not match the catalog case', observed: `got ${preview(observed)}, want ${preview(c.decoded)}` };
  }
  return { ...base, outcome: 'pass', observed: preview(observed) };
}

function main() {
  const args = parseArgs(process.argv);
  const suites = loadSuites().filter((s) => !args.formats.length || args.formats.includes(s.format));
  const adapters = makeAdapters().filter((a) => !args.serializers.length || args.serializers.includes(a.name));
  const byFormat = new Map();
  for (const a of adapters) {
    if (!byFormat.has(a.format)) byFormat.set(a.format, []);
    byFormat.get(a.format).push(a);
  }
  const results = [];
  const adapterErrors = [];
  for (const suite of suites) {
    const chosen = suite.adapters?.length
      ? (byFormat.get(suite.format) || []).filter((a) => suite.adapters.includes(a.name))
      : byFormat.get(suite.format) || [];
    if (!chosen.length) {
      adapterErrors.push(`No adapter registered for format ${suite.format} (${suite.standard} (${suite.version}))`);
      continue;
    }
    for (const adapter of chosen) {
      for (const c of suite.cases) {
        results.push(runOne(suite, c, adapter));
      }
    }
  }
  const passed = results.filter((r) => r.outcome === 'pass').length;
  const failed = results.filter((r) => r.outcome === 'fail').length;
  const skipped = results.filter((r) => r.outcome === 'skip').length;
  const errors = results.filter((r) => r.outcome === 'error').length;
  const by = new Map();
  for (const r of results) {
    const key = `${r.standard} (${r.version}) × ${r.serializer}`;
    if (!by.has(key)) by.set(key, { p: 0, f: 0, s: 0 });
    const cell = by.get(key);
    if (r.outcome === 'pass') cell.p += 1;
    else if (r.outcome === 'fail') cell.f += 1;
    else cell.s += 1;
  }
  console.log('Serialization compliance (library deviations are catalogued, not a red build)');
  console.log(`  ${passed} pass  ${failed} fail  ${skipped} skip  ${errors} error  ${results.length} total`);
  if (adapterErrors.length) {
    console.log('  Adapter errors:');
    for (const e of adapterErrors) console.log(`    - ${e}`);
  }
  console.log('  By suite × adapter:');
  for (const [k, v] of [...by.entries()].sort()) {
    console.log(`    ${k}: ${v.p} pass, ${v.f} fail, ${v.s} skip`);
  }
  const report = {
    schema: 'gld.dashboard.compliance/1',
    generated_at: new Date().toISOString().replace(/\.\d+Z$/, 'Z'),
    language: 'javascript',
    languages: ['javascript'],
    policy: 'report-only',
    scope: { formats: [...new Set(results.map((r) => r.format))].sort() },
    passed,
    failed,
    skipped,
    errors,
    catalog_errors: [],
    serializer_errors: adapterErrors,
    results,
  };
  if (args.jsonOut) {
    const dest = resolve(args.jsonOut);
    mkdirSync(dirname(dest), { recursive: true });
    writeFileSync(dest, `${JSON.stringify(report, null, 2)}\n`);
    console.log(`\nWrote ${dest}`);
  }
}

main();
